### ISSUE TITLE
openHAB sometimes doesn't start properly after upgrades

### ISSUE THREAD

**Issue by mstormi** [openhab/openhab-distro#1696]
I noticed on various occasions that openHAB didn't start properly after a package upgrade.

On every of these, I had to manually clean the cache to make it finally start (or to proceed after start, actually - it starts in fact but keeps hanging).

@BClark09 I'd prefer if we sort this out before OH 4.3 release Dec 15.

**Comment by BClark09** [comment]
When you mention Karaf cache, do you mean `cache` and `tmp` directories inside `${OPENHAB_USERDIR}`? or just the `karaf` dir?

If `cache` and `tmp`, these are indeed removed in the [distro's update scripts](https://github.com/openhab/openhab-distro/blob/main/distributions/openhab/src/main/resources/bin/update#L321C1-L324C43) without condition. (Note that [Linuxpkg does this in a similar way](https://github.com/openhab/openhab-linuxpkg/blob/main/resources/control-runtime/preinst#L84). Have you tried completing an upgrade without openHAB running and then looking to see if the files still exist?

If `.karaf`, what is stored here? I thought it was just the console history?

**Comment by mstormi** [comment]
did not manually remove anything. Used `openhab-cli clean-cache`, can you compare what that does to those dirs you named?

Am 14. November 2024 19:40:44 MEZ schrieb Ben Clark ***@***.***>:
>When you mention Karaf cache, do you mean `cache` and `tmp` directories inside `${OPENHAB_USERDIR}`? or just the `karaf` dir?
>
>If `cache` and `tmp`, these are indeed removed in the [distro's update scripts](https://github.com/openhab/openhab-distro/blob/main/distributions/openhab/src/main/resources/bin/update#L321C1-L324C43) without condition. (Note that [Linuxpkg does this in a similar way](https://github.com/openhab/openhab-linuxpkg/blob/main/resources/control-runtime/preinst#L84)). Have you tried completing an upgrade without openHAB running and then looking to see if the files still exist?
>
>If `.karaf`, what is stored here? I thought it was just the console history?
>

**Comment by mike-bike** [comment]
Not sure if it helps. I do see start-up issues (Bindings not all initializing, etc) on the first start after an upgrade. Most of the time I do a stop, clean-cache, restart. However, I was also able to get a clean start after a stop and start without cache clean. Maybe there are sometimes timing issues in the **first** start after an upgrade. Unfortunately, this seems not to be reproducible.

**Comment by andrewfg** [comment]
> Most of the time I do a stop, clean-cache, restart

Me too. This should be automated!

**Comment by andrewfg** [comment]
I suppose that if the cached files are in memory then deleting the disk images does not automatically clear the memory image. Also probably a service restart won't clear the memory images either. I suppose that only a full shutdown and reboot would do that.

**Comment by BClark09** [comment]
The openHAB service is shutdown during an upgrade, so there should be no instance in memory.

**Comment by mstormi** [comment]
> You can see from above that the `cache` and `tmp` directories have indeed been cleaned during the upgrade process. This will be the same if you're using the distributions update scripts.

Did you check if exactly the same will happen when I manually run `openhab-cli clean-cache` ?

**Comment by BClark09** [comment]
I only have my real-home system at the moment, so won't manually run the command myself. But as far as I can see there's no functional difference in the commands it uses vs the update scripts:

https://github.com/openhab/openhab-linuxpkg/blob/main/resources/usr/bin/openhab-cli#L99-L101

**Comment by andrewfg** [comment]
> When you mention Karaf cache, do you mean cache and tmp directories

From your script above it looks like the update script also cleans /marketplace as well as /tmp and /cache directories.

**Comment by BClark09** [comment]
Yep, this was necessary because of https://github.com/openhab/openhab-core/pull/4019

**Comment by wborn** [comment]
> I noticed on various occasions that openHAB didn't start properly after a package upgrade.

Please add some logging etc. of when this happens. If you can reproduce it maybe you can also add debug logging of openhab/karaf as well.

**Comment by mstormi** [comment]
While I currently don't have a handle to tackle the root cause(s), I wonder if we can improve the situation.

The first start after an upgrade/clean-cache quite often results in Java spitting gazillions of messages in some sort of loop
(have been unable to identify the reason for it).

I then have to kill -9 the java process to get out of this loop.

The standard openhab start script `/usr/share/openhab/runtime/bin/karaf` apparently does not handle that situation well enough.

@BClark09 Wouldn't it make sense to build a wrapper that auto-kills Java in these cases?
